File: src/leetCode/c++/Solution199.cpp
//BFS O(V) O(V)
class Solution199 {
public:
    vector<int> rightSideView(TreeNode* root) {
        if(!root) return {};

        vector<int> res;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            int num = -1;
            for(int i=q.size(); i>0; --i){
                TreeNode* node = q.front(); q.pop();
                num = node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
            res.push_back(num);
        }
        return res;
    }
};

File: java/Solution199.java
package leetCode.java;

import java.util.*;

//BFS O(V) O(V)
class Solution199 {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }

        Queue<TreeNode> q = new ArrayDeque<>();
        q.offer(root);
        while (!q.isEmpty()) {
            int rightVal = -1;
            for (int i = q.size(); i > 0; --i) {
                TreeNode node = q.poll();
                rightVal = node.val;
                if (node.left != null) {
                    q.offer(node.left);
                }
                if (node.right != null) {
                    q.offer(node.right);
                }
            }
            res.add(rightVal);
        }
        return res;
    }
}